File: android/src/main/kotlin/com/squareup/square_mobile_payments_sdk/modules/AuthModule.kt
package com.squareup.square_mobile_payments_sdk.modules

import com.squareup.sdk.mobilepayments.MobilePaymentsSdk
import com.squareup.sdk.mobilepayments.core.Result as SdkResult
import com.squareup.square_mobile_payments_sdk.extensions.toAuthorizationStateName
import com.squareup.square_mobile_payments_sdk.extensions.toAuthorizeErrorCodeName
import com.squareup.square_mobile_payments_sdk.extensions.toAuthorizedLocationMap
import com.squareup.square_mobile_payments_sdk.extensions.toErrorDetailsMap
import io.flutter.plugin.common.MethodChannel

class AuthModule {
  companion object {
    private val authManager = MobilePaymentsSdk.authorizationManager()

    @JvmStatic
    fun getAuthorizationState(result: MethodChannel.Result) {
      result.success(authManager.authorizationState.toAuthorizationStateName())
    }

    @JvmStatic
    fun getAuthorizedLocation(result: MethodChannel.Result) {
      val location = authManager.location
      if (location == null) {
        result.success(null)
        return
      }
      result.success(location.toAuthorizedLocationMap())
    }

    @JvmStatic
    fun authorize(result: MethodChannel.Result, accessToken: String?, locationId: String?) {
      if (accessToken == null || locationId == null) {
        result.error(
          "missingParameters",
          "Missing accessToken or locationId",
          null
        )
        return
      }
      authManager.authorize(accessToken, locationId) { sdkResult ->
        when (sdkResult) {
          is SdkResult.Success -> {
            result.success(null)
          }
          is SdkResult.Failure -> {
            result.error(
                    sdkResult.errorCode.toAuthorizeErrorCodeName(),
                    sdkResult.errorMessage,
                    sdkResult.details.map { d -> d.toErrorDetailsMap() }
            )
          }
          else -> {
            result.error("unknown", null, null)
          }
        }
      }
    }

    @JvmStatic
    fun deAuthorize(result: MethodChannel.Result) {
      authManager.deauthorize()
      result.success(null)
    }
  }
}

File: ios/Classes/Modules/AuthModule.swift
import Flutter
import SquareMobilePaymentsSDK

public class AuthModule {
    private static let authManager = MobilePaymentsSDK.shared.authorizationManager

    public static func getAuthorizationState(result: @escaping FlutterResult) {
        return result(authManager.state.getName())
    }

    public static func getAuthorizedLocation(result: @escaping FlutterResult) {
        guard let location = authManager.location else {
            return result(nil)
        }
        return result(location.toMap())
    }

    public static func authorize(
        result: @escaping FlutterResult,
        accessToken: String,
        locationId: String) {
            authManager.authorize(
                withAccessToken: accessToken,
                locationID: locationId
            ) { error in
                if let error {
                    var e = error as NSError
                    if let authError = AuthorizationError(rawValue: e.code) {
                        result(FlutterError(code: authError.getName(),
                                        message: e.localizedDescription,
                                        details: e.localizedFailureReason))
                    }
                } else {
                    result(NSNull())
                }
            }
    }

    public static func deauthorize(result: @escaping FlutterResult) {
        authManager.deauthorize {
            result(NSNull())
        }
    }
}